### ISSUE TITLE
[wxGTK] wxStyledTextCtrl - bug when using HiDPI

### ISSUE THREAD

**Issue by eranif** [wxWidgets/wxWidgets#22512]
**Describe the bug**
When enabling HiDPI or fractional text scaling (for example: 125%), there is an unwanted, unclickable margin which makes the
edit control unusable.
Beside the appearance disturbance, it also causes everything to "shift" to the right.
e.g. if I click with the mouse on column 1, the caret will appear in column 3 or 4

(image attached)
![stc-bug](https://user-images.githubusercontent.com/3677711/173081217-77004ceb-92e4-4ad1-9882-b70671e54c8f.png)

**Expected vs observed behaviour**
See above bug report

**Patch or snippet allowing to reproduce the problem**
To reproduce the problem:

- Open the stc sample
- Open the file `edit.cpp` and add somewhere in the function `Edit::InitializePrefs` this line `SetBufferedDraw(true);` (I added it at line 616)
- Re-compile the sample
- Start the sample, make sure that HiDPI is set

To enable HiDPI, you can either do that from the UI, or from the command line like this:

```bash
export GDK_SCALE=2
./stctest
```
**Platform and version information**
 - wxWidgets version you use: 3.16
 - wxWidgets port you use: wxGTK
 - Ubuntu 20.04

The simple fix (for me) was to simply remove the `SetBufferedDraw(true)` from the code...

This bug was reported by multiple users of mine:

https://github.com/eranif/codelite/issues/2992
https://github.com/eranif/codelite/issues/3010

**Comment by MaartenBent** [comment]
Something inside Scintilla (or wxSTC) doesn't seem to handle the margin width correctly. When drawing each line from the buffer onto the screen, the wrong origin point is used.
For example, with the 16px margin in the STC sample, at 2x scaling it uses `x=16` instead of `x=32` as the origin point. This causing all the lines to shift 16px to the right. The void is filled with black pixels. You have line numbers enabled, so the gap is bigger.

I don't notice any issues on macOS. Maybe the problem could be in the GTK `Blit` implementation? 

An awful hack that seems to fix it for me is adjusting the origin point when blitting the image onto the screen:
```diff
diff --git "a/src/stc/PlatWX.cpp" "b/src/stc/PlatWX.cpp"
index 61d3551bdf..0b0a309007 100644
--- "a/src/stc/PlatWX.cpp"
+++ "b/src/stc/PlatWX.cpp"
@@ -559,9 +559,14 @@ void SurfaceImpl::Ellipse(PRectangle rc, ColourDesired fore, ColourDesired back)
 
 void SurfaceImpl::Copy(PRectangle rc, Point from, Surface &surfaceSource) {
     wxRect r = wxRectFromPRectangle(rc);
+    wxPoint f(from.x, from.y);
+#ifdef __WXGTK__
+    f = f * hdc->GetContentScaleFactor();
+#endif
+
     hdc->Blit(r.x, r.y, r.width, r.height,
               ((SurfaceImpl&)surfaceSource).hdc,
-              wxRound(from.x), wxRound(from.y), wxCOPY);
+              f.x, f.y, wxCOPY);
 }
 
 void SurfaceImpl::DrawTextNoClip(PRectangle rc, Font &font, XYPOSITION ybase,
```

**Comment by eranif** [comment]
Do we need `Buffered Draw` on GTK? I did not notice any flicker when disabling it...
so I wonder whether we can simply make it a function that does nothing under GTK

Did you notice any difference under GTK?

**Comment by vadz** [comment]
There were problems with blit origin in wxGTK before, but I thought they were fixed a long time ago, see `wxHAS_WORKING_GTK_DC_BLIT` in `include/wx/gtk/dc.h`.

If there are still problems, it would be great to reproduce them in a simple example or, ideally, a unit test, and either fix `Blit()` itself or at least document its limitations, so that people could work around it in their code.

Maarten, are you already looking at this? If yes, I'd rather spend time on other things, but if not, please let me know as this seems to be a rather bad problem and I'd like to fix it before 3.2.0 if possible. TIA!

**Comment by MaartenBent** [comment]
I don't notice differences on GTK. It is supposed to have native buffering so I don't think you need it. `wxAutoBufferedPaintDC` doesn't use a buffer either:
https://github.com/wxWidgets/wxWidgets/blob/e7785ac224f59909e2f3b0c14cf62b4af71ed282/include/wx/dcbuffer.h#L20-L24

It doesn't seem related to `wxHAS_WORKING_GTK_DC_BLIT` which is only used when scrolling in univ build.

I did some tests in the image sample, and I suspect it might be related to `wxBitmap::GetSubBitmap` that is used when blitting (via `wxGCDCImpl::DoStretchBlit`, `source->GetAsBitmap`).

I modify the image sample like this:
```diff
diff --git "a/samples/image/canvas.cpp" "b/samples/image/canvas.cpp"
index 9a634fed4f..fd1a642e3c 100644
--- "a/samples/image/canvas.cpp"
+++ "b/samples/image/canvas.cpp"
@@ -416,11 +416,25 @@ void MyCanvas::OnPaint( wxPaintEvent &WXUNUSED(event) )
         dc.DrawBitmap( my_anti, 280, 30 );
 
     dc.DrawText( "PNG handler", 30, 135 );
-    if (my_horse_png.IsOk())
-    {
-        dc.DrawBitmap( my_horse_png, 30, 150 );
-        wxRect rect(0,0,100,100);
-        wxBitmap sub( my_horse_png.GetSubBitmap(rect) );
+    wxBitmap test;
+    test.CreateWithDIPSize(100, 100, dc.GetContentScaleFactor());
+    {
+        wxMemoryDC memdc;
+        memdc.SelectObject(test);
+        memdc.SetBrush(*wxRED_BRUSH);
+        memdc.DrawRectangle(0, 0, FromDIP(50), FromDIP(50));
+        memdc.SetBrush(*wxGREEN_BRUSH);
+        memdc.DrawRectangle(0, FromDIP(50), FromDIP(50), FromDIP(50));
+        memdc.SetBrush(*wxBLUE_BRUSH);
+        memdc.DrawRectangle(FromDIP(50), 0, FromDIP(50), FromDIP(50));
+        memdc.SetBrush(*wxYELLOW_BRUSH);
+        memdc.DrawRectangle(FromDIP(50), FromDIP(50), FromDIP(50), FromDIP(50));
+    }
+    if (test.IsOk())
+    {
+        dc.DrawBitmap(test, 30, 150 );
+        wxRect rect(FromDIP(25), FromDIP(25), FromDIP(50), FromDIP(50));
+        wxBitmap sub(test.GetSubBitmap(rect) );
         dc.DrawText( "GetSubBitmap()", 280, 175 );
         dc.DrawBitmap( sub, 280, 195 );
     }
```

On Windows 10 150% it looks like:
![windows](https://user-images.githubusercontent.com/8088070/173195759-319184fa-165d-435d-889f-bb0f623b4039.png)

But On Ubuntu 21.10 GTK3 200% it looks like:
![Screenshot from 2022-06-11 18-06-02](https://user-images.githubusercontent.com/8088070/173195766-a66133f9-5753-458e-8f97-e92da5e08c74.png)

MacOS 11.6 retina is fine too:
![macos](https://user-images.githubusercontent.com/8088070/173195800-606d377f-6f10-45d7-acd7-bbc64aa4a305.png)

**Comment by MaartenBent** [comment]
Oh, it doesn't use `wxGCDCImpl::DoStretchBlit` but `wxGTKCairoDCImpl::DoStretchBlit`. But maybe it has a similar problem. I'll continue investigating.

**Comment by MaartenBent** [comment]
With this patch the problem in wxSTC is fixed:
```diff
diff --git "a/src/gtk/dc.cpp" "b/src/gtk/dc.cpp"
index 574497437d..9fc38a01b2 100644
--- "a/src/gtk/dc.cpp"
+++ "b/src/gtk/dc.cpp"
@@ -310,7 +310,7 @@ bool wxGTKCairoDCImpl::DoStretchBlit(int xdest, int ydest, int dstWidth, int dst
     const double bmpScale = bitmap.IsOk() ? bitmap.GetScaleFactor() : 1.0;
 
     cairo_scale(cr, dstWidth / (sx * srcWidth * bmpScale), dstHeight / (sy * srcHeight * bmpScale));
-    cairo_set_source_surface(cr, surfaceSrc, -xsrc_dev, -ysrc_dev);
+    cairo_set_source_surface(cr, surfaceSrc, -xsrc_dev * bmpScale, -ysrc_dev * bmpScale);
     const wxRasterOperationMode rop_save = m_logicalFunction;
     SetLogicalFunction(rop);
     cairo_pattern_set_filter(cairo_get_source(cr), CAIRO_FILTER_NEAREST);
```

Is anyone more familiar with this code or cairo/gtk/blitting to know if this is correct? 
And `cairo_mask_surface` below would probably need the same fix.

**Comment by vadz** [comment]
Thanks a lot for finding this!

I'm afraid the only person really familiar with this code must be @paulcor, but the fix makes sense to me: `[xy]src_dev` don't depend on the scaling factor, but clearly the resulting offset must depend on it, so it seems like it should be applied here.

**Comment by paulcor** [comment]
Should be fixed by 8e4a4c8dc6
